src/data/tprod.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2020 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import data.list.nodup
+
+/-!
+# Finite Products of Types
+
+This file defines the product of types over a list. For `l : list ι` and `α : ι → Type*` we define
+`tprod α l = l.foldr (λ i β, α i × β) punit`. It is used to transfer results from binary products to
+finitary products, like for product measures.
+
+## Main definitions
+
+* We have the equivalence `tprod.pi_equiv_tprod : (Π i, α i) ≃ tprod α l`
+  if `l` contains every element of `ι` exactly once.
+* The product of sets is `set.tprod : (Π i, set (α i)) → set (tprod α l)`.
+-/
+
+open list function
+
+variables {ι : Type*} (α : ι → Type*) {i j : ι} {l : list ι} {f : Π i, α i}
+
+/-- The product of a family of types over a list. -/
+@[nolint has_inhabited_instance] def tprod (l : list ι) : Type* :=
+l.foldr (λ i β, α i × β) punit
+
+variable {α}
+
+namespace tprod
+
+open list
+
+/-- Turning a function `f : Π i, α i` into an element of the iterated product `tprod α l`. -/
+protected def mk : ∀ (l : list ι) (f : Π i, α i), tprod α l
+| []        := λ f, punit.star
+| (i :: is) := λ f, (f i, mk is f)
```

Comment:
Is this the same as `l.foldr (λ i x, (f i, x)) punit.star`? Is there a reason to not use that spelling?

Reply:
No, `list.foldr` requires that it be nondependent, but this definition has a type that depends on `l`.

Reply:
Does it make sense to generalize this to `list. dfoldl` for things like iterated `sum` types too?

Reply:
Eg, something like:
```

def list.dfoldr
  {α : Type*}
  {β : α → Type* → Type*}
  {γ : Type*}
  (f : Π {T : Type*} (a : α), T → β a T) (init : γ) : Π (l : list α), l.foldr β γ
| [] := init
| (a :: as) := f a as.dfoldr

/-- Turning a function `f : Π i, α i` into an element of the iterated product `tprod α l`. -/
protected def mk' (l : list ι) (f : Π i, α i) : tprod α l :=
l.dfoldr (λ _ i b, (f i, b)) punit.star
```

Reply:
I'm not sure I see the advantage. Yes you can do this, in the same sense that any recursive function is expressible in terms of `list.rec`, but that doesn't make it easier to work with. The equation compiler definition is both more clear and also automatically generates the appropriate definitional lemmas.